k2pow-service/src/main.rs

```diff
@@ -0,0 +1,377 @@
+    _ = tokio::task::spawn(consume(
+        job_manager.clone(),
+        args.cores,
+        args.randomx_mode,
+        args.randomx_large_pages,
+    ));
+    let router = router(job_manager);
+    tracing::info!(
+        "starting http server with bind address: {}",
+        args.bind_address
+    );
+    let listener = tokio::net::TcpListener::bind(args.bind_address)
+        .await
+        .unwrap();
+    axum::serve(listener, router).await.unwrap();
+
+    Ok(())
+}
+
+fn router<T: GetOrCreate>(job_manager: Arc<T>) -> Router {
+    Router::new()
+        .route("/", get(root))
+        .route(
+            "/job/:miner/:nonce_group/:challenge/:difficulty",
+            get(get_job),
+        )
+        .with_state(job_manager)
+        .layer(
+            TraceLayer::new_for_http()
+                .make_span_with(|request: &Request<_>| {
+                    let matched_path = request.uri().to_string();
+
+                    info_span!(
+                        "http_request",
+                        method = ?request.method(),
+                        uri =  matched_path,
+                        status = tracing::field::Empty,
+                    )
+                })
+                .on_response(|response: &Response, _latency: Duration, span: &Span| {
+                    span.record("status", response.status().as_str());
+                    tracing::info!("served request");
+                })
+                .on_failure(
+                    |error: ServerErrorsFailureClass, _latency: Duration, span: &Span| {
+                        match error {
+                            ServerErrorsFailureClass::StatusCode(code) => {
+                                span.record("status", code.as_str());
+                            }
+                            ServerErrorsFailureClass::Error(err) => {
+                                span.record("error", &err);
+                            }
+                        }
+                        tracing::error!("request fail");
+                    },
+                ),
+        )
+}
+async fn consume(
+    job_manager: Arc<JobManager>,
+    cores: u8,
+    randomx_mode: RandomXMode,
+    randomx_large_pages: bool,
+) {
+    let mut randomx_flags = match randomx_mode {
+        RandomXMode::Fast => {
+            post::pow::randomx::RandomXFlag::get_recommended_flags()
+                | post::pow::randomx::RandomXFlag::FLAG_FULL_MEM
+        }
+        RandomXMode::Light => post::pow::randomx::RandomXFlag::get_recommended_flags(),
+    };
+    if randomx_large_pages {
+        eprintln!("Using large pages for RandomX");
+        randomx_flags |= post::pow::randomx::RandomXFlag::FLAG_LARGE_PAGES;
+    }
+    eprintln!("RandomX flags: {}", randomx_flags);
+
+    loop {
+        sleep(Duration::from_secs(POLL_EVERY_SECS)).await;
+        let job = match job_manager.take().await {
+            None => continue,
+            Some(v) => v,
+        };
```

Comment:
Yeah so I don't mind doing it this way, but I'd chuck away and throw the whole job manager with it if I'd do so because it doesn't make much sense in this case to have it. If the job manager already returns a handle then it's already coupled enough to just have the axum handler to take a `HashMap` as `State`, and just throw everything else to the dustbin.

> You would avoid having another thread just for starting the tasks

It's not a separate thread, but just an async task so I'm not sure the overhead is so big.

Reply:
> If the job manager already returns a handle then it's already coupled enough to just have the axum handler to take a HashMap as State, and just throw everything else to the dustbin.

It doesn't need to return the handle. It might have an enum for its inner logic (keeping job's state), which would have the additional data like the job handle and a different one returned to the caller (this one would only say "in progress" vs "done, here's the result (or error)". Translation from the former to the latter is trivial.

> > You would avoid having another thread just for starting the tasks
>
> It's not a separate thread, but just an async task so I'm not sure the overhead is so big.

It's not about the performance but the unnecessary complexity of the whole thing. Btw, IMHO async is worse because the code gets more complicated quickly. It would be great if JobManager could be "sync".

Reply:
we can do it this way yeah. I don't mind doing the change. All in all, I'm not sure how much code it will really save on EOD... But it's fine to do it this way :+1: